File: advance/src/eu/advance/logistics/flow/engine/model/fd/AdvanceType.java
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */

package eu.advance.logistics.flow.engine.model.fd;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.Arrays;
import java.util.List;

import com.google.common.collect.Iterables;
import com.google.common.collect.Lists;

import edu.umd.cs.findbugs.annotations.NonNull;
import eu.advance.logistics.flow.engine.inference.Type;
import eu.advance.logistics.flow.engine.inference.TypeKind;
import eu.advance.logistics.flow.engine.typesystem.XType;
import eu.advance.logistics.flow.engine.xml.XElement;
import eu.advance.logistics.flow.engine.xml.XSerializable;

/**
 * The type parameter bound definition for a generic type parameter.
 * @author akarnokd, 2011.07.01.
 */
public class AdvanceType implements Type, XSerializable {
	/** The counter used for type variables. */
	public int id;
	/** Reference to another type parameter within the same set of declarations. */
	public String typeVariableName;
	/** The actual type variable object. */
	public AdvanceTypeVariable typeVariable;
	/** An existing concrete type definition schema. */
	public URI typeURI;
	/** The concrete XML type of the target schema. */
	public XType type;
	/** The type arguments used by the concrete type. */
	public final List<AdvanceType> typeArguments = Lists.newArrayList();
	@Override
	public String toString() {
		if (typeURI != null) {
			if (typeArguments.isEmpty()) {
				return typeURI.toString();
			}
			StringBuilder b = new StringBuilder(typeURI.toString());
			b.append("<");
			int i = 0;
			for (AdvanceType ta : typeArguments) {
				if (i > 0) {
					b.append(", ");
				}
				b.append(ta);
				i++;
			}
			b.append(">");
			return b.toString();
		}
		if (id > 0) {
			return String.format("%s[%d]", typeVariableName, id);
		}
		return typeVariableName;
		
	}
	/** @return create a new instance of this type declaration. */
	public AdvanceType copy() {
		AdvanceType result = new AdvanceType();
		result.id = id;
		result.typeVariableName = typeVariableName;
		result.typeURI = typeURI;
		result.type = type;
		if (typeVariable != null) {
			result.typeVariable = typeVariable.copy();
		}
		for (AdvanceType ta : typeArguments) {
			result.typeArguments.add(ta.copy());
		}
		return result;
	}
	/**
	 * Create a fresh type variable with the given name.
	 * @param name the type variable name
	 * @return Construct a fresh type variable with name.
	 */
	public static AdvanceType fresh(String name) {
		AdvanceType result = new AdvanceType();
		result.typeVariableName = name;
		result.typeVariable = new AdvanceTypeVariable();
		result.typeVariable.name = name;
		return result;
	}
	/**
	 * Load a type description from an XML element which conforms the {@code block-description.xsd}.
	 * @param root the root element of an input/output node.
	 */
	@Override
	public void load(XElement root) {
		typeVariableName = root.get("type-variable");
		String tu = root.get("type");
		if ((tu != null) == (typeVariableName != null)) {
			throw new IllegalArgumentException("Only one of the type-variable and type arguments should be defined! " + root);
		}
		if (tu != null) {
			try {
				typeURI = new URI(tu);
			} catch (URISyntaxException ex) {
				throw new RuntimeException(ex);
			}
		}
		id = root.getInt("type-id", 0);
		for (XElement ta : root.childrenWithName("type-argument")) {
			AdvanceType at = new AdvanceType();
			at.load(ta);
			typeArguments.add(at);
		}
	}
	@Override
	public void save(XElement destination) {
		if (id > 0) {
			destination.set("type-id", id);
		}
		destination.set("type-variable", typeVariableName);
		destination.set("type", typeURI);
		for (AdvanceType at : typeArguments) {
			at.save(destination.add("type-argument"));
		}
	}
	@Override
	public TypeKind kind() {
		if (typeURI != null) {
			if (typeArguments.isEmpty()) {
				return TypeKind.CONCRETE_TYPE;
			}
			return TypeKind.PARAMETRIC_TYPE;
		}
		return TypeKind.VARIABLE_TYPE;
	}
	/**
	 * Creates a concrete (e.g., no type arguments or parametric type with the supplied base type and type arguments.
	 * <p>Note: the {@code AdvanceType.type} is not resolved.</p>
	 * <p>Example:</p>
	 * <pre>createType(COLLECTION, createType(STRING))</pre>
	 * @param baseType the base type
	 * @param types the type arguments.
	 * @return the created new type
	 */
	@NonNull
	public static AdvanceType createType(@NonNull URI baseType, AdvanceType... types) {
		return createType(baseType, Arrays.asList(types));
	}
	/**
	 * Creates an unbounded variable type.
	 * @param name the variable name
	 * @return the created type
	 */
	@NonNull
	public static AdvanceType createType(@NonNull String name) {
		AdvanceType t = new AdvanceType();
		t.typeVariableName = name;
		t.typeVariable = new AdvanceTypeVariable();
		t.typeVariable.name = name;
		return t;
	}
	/**
	 * Creates a bounded variable type.
	 * @param name the variable name
	 * @param upperBound defines if the type bounds are the upper bounds, e.g., {@code T super B1, B2, B3}.
	 * @param bound1 the first bound
	 * @param boundsRest the rest of the bounds
	 * @return the created type
	 */
	@NonNull
	public static AdvanceType createType(@NonNull String name, boolean upperBound, AdvanceType bound1, AdvanceType... boundsRest) {
		AdvanceType t = new AdvanceType();
		t.typeVariableName = name;
		t.typeVariable = new AdvanceTypeVariable();
		t.typeVariable.name = name;
		t.typeVariable.isUpperBound = upperBound;
		t.typeVariable.bounds.add(bound1);
		t.typeVariable.bounds.addAll(Arrays.asList(boundsRest));
		return t;
	}
	/**
	 * Creates a bounded variable type.
	 * @param name the variable name
	 * @param upperBound defines if the type bounds are the upper bounds, e.g., {@code T super B1, B2, B3}.
	 * @param bounds the type bounds
	 * @return the created type
	 */
	@NonNull
	public static AdvanceType createType(@NonNull String name, boolean upperBound, Iterable<? extends AdvanceType> bounds) {
		AdvanceType t = new AdvanceType();
		t.typeVariableName = name;
		t.typeVariable = new AdvanceTypeVariable();
		t.typeVariable.name = name;
		t.typeVariable.isUpperBound = upperBound;
		Iterables.addAll(t.typeVariable.bounds, bounds);
		return t;
	}
	
	/**
	 * Creates a concrete (e.g., no type arguments or parametric type with the supplied base type and type arguments.
	 * <p>Note: the {@code AdvanceType.type} is not resolved.</p>
	 * <p>Example:</p>
	 * <pre>createType(COLLECTION, createType(STRING))</pre>
	 * @param baseType the base type
	 * @param types the type arguments.
	 * @return the created new type
	 */
	@NonNull
	public static AdvanceType createType(@NonNull URI baseType, Iterable<? extends AdvanceType> types) {
		AdvanceType t = new AdvanceType();
		t.typeURI = baseType;
		Iterables.addAll(t.typeArguments, types);
		return t;
		
	}
	/**
	 * Creates an {@code advance:type} type constructor XML.
	 * @param type the type to convert to XML
	 * @return the XML representing the type constructor
	 */
	public static XElement createType(AdvanceType type) {
		XElement result = new XElement("type");
		type.save(result);
		return result;
	}
	/**
	 * Converts the XML type declaration into an actual AdvanceType object.
	 * @param type the type XML
	 * @return the  the type object
	 */
	public static AdvanceType getType(XElement type) {
		AdvanceType t = new AdvanceType();
		t.load(type);
		return t;
	}
}

File: advance-api-cs/AdvanceAPIClient/Classes/Model/AdvanceType.cs
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Xml;

using AdvanceAPIClient.Core;
using AdvanceAPIClient.Classes.Typesystem;
using AdvanceAPIClient.Interfaces;

namespace AdvanceAPIClient.Classes.Model
{
    /// <summary>
    /// Type parameter bound definition for a generic type parameter
    /// </summary>
    public class AdvanceType : XmlReadWrite, IType
    {
        /// <summary>
        /// Counter used for type variables
        /// </summary>
        public int Id;
        /// <summary>
        /// Reference to another type parameter within the same set of declarations
        /// </summary>
        public string TypeVariableName;
        /// <summary>
        /// actual type variable object
        /// </summary>
        public AdvanceTypeVariable TypeVariable;
        /// <summary>
        /// Existing type definition schema 
        /// </summary>
        public Uri TypeURI;
        /// <summary>
        /// XML type of the target schema
        /// </summary>
        public XType Type;
        /// <summary>
        /// Type arguments used by the type
        /// </summary>
        /// <typeparam name="AdvanceType"></typeparam>
        /// <param name="?"></param>
        /// <returns></returns>
        public List<AdvanceType> TypeArguments = new List<AdvanceType>();

        public TypeKind Kind
        {
            get
            {
                if (this.TypeURI == null)
                    return TypeKind.VARIABLE_TYPE;
                else if (this.TypeArguments.Count == 0)
                    return TypeKind.CONCRETE_TYPE;
                else
                    return TypeKind.PARAMETRIC_TYPE;
            }
        }


        public AdvanceType() : base() { }

        public override string ToString()
        {
            if (this.TypeURI != null)
            {
                if (this.TypeArguments.Count == 0)
                    return this.TypeURI.ToString();
                else
                {
                    StringBuilder b = new StringBuilder(this.TypeURI.ToString());
                    b.Append("<");
                    int i = 0;
                    foreach (AdvanceType ta in this.TypeArguments)
                    {
                        if (i > 0) b.Append(", ");
                        b.Append(ta);
                        i++;
                    }
                    b.Append(">");
                    return b.ToString();
                }
            }
            else if (this.Id > 0)
                return string.Format("{0}[{1}]", this.TypeVariableName, this.Id);
            else
                return this.TypeVariableName;
        }

        /// <summary>
        /// Creates new instance of this type
        /// </summary>
        /// <returns>new instance of this type declaration</returns>
        public AdvanceType Copy()
        {
            AdvanceType result = new AdvanceType();
            result.Id = this.Id;
            result.TypeVariableName = this.TypeVariableName;
            result.TypeURI = this.TypeURI;
            result.Type = Type;
            if (this.TypeVariable != null)
                result.TypeVariable = this.TypeVariable.Copy();
            foreach (AdvanceType ta in this.TypeArguments)
            {
                result.TypeArguments.Add(ta.Copy());
            }
            return result;
        }

        /// <summary>
        /// Create a fresh type variable with given name.
        /// </summary>
        /// <param name="name">Type variable name</param>
        /// <returns>type variable</returns>
        public static AdvanceType Fresh(string name)
        {
            return CreateType(name);
        }

        /// <summary>
        /// Fill object data from xml node
        /// </summary>
        /// <param name="source"></param>
        protected override void LoadFromXmlNode(XmlNode source)
        {
            this.Id = GetIntAttribute(source, "type-id", 0);
            this.TypeVariableName = GetAttribute(source, "type-variable", null);
            this.TypeURI = GetUriAttribute(source, "type");
            if ((this.TypeURI != null) == (this.TypeVariableName != null))
                this.ThrowException("Only one of the type-variable and type arguments should be defined! " + source.InnerText);
            foreach (XmlNode ta in GetChildren(source, "type-argument"))
                this.TypeArguments.Add(CreateFromXml<AdvanceType>(ta));
        }

        /// <summary>
        /// Fills object's xml
        /// </summary>
        /// <param name="node">Empty node for the object with name "tagname"</param>
        protected override void FillXmlElement(XmlElement node)
        {
            if (this.Id > 0)
                AddAttribute(node, "type-id", this.Id);
            AddAttribute(node, "type-variable", this.TypeVariableName);
            if (this.TypeURI != null)
                AddAttribute(node, "type", this.TypeURI);
            foreach (AdvanceType at in this.TypeArguments)
                at.AddToXML("type-argument", node);
        }

        /// <summary>
        /// Creates a concrete (e.g., no type arguments or parametric type 
        /// with the supplied base type and type arguments.
        /// </summary>
        /// <remarks>
        /// {@code AdvanceType.type} is not resolved.
        /// </remarks>
        /// <example>
        /// CreateType(COLLECTION, CreateType(STRING))
        /// </example>
        /// <param name="baseType">Base type</param>
        /// <param name="types">type arguments</param>
        /// <returns>created type</returns>
        public static AdvanceType CreateType(Uri baseType, params AdvanceType[] types)
        {
            return CreateType(baseType, new List<AdvanceType>(types));
        }

        /// <summary>
        /// Creates an unbounded variable type.
        /// </summary>
        /// <param name="name">variable name</param>
        /// <returns>created type</returns>
        public static AdvanceType CreateType(string name)
        {
            AdvanceType t = new AdvanceType();
            t.TypeVariableName = name;
            t.TypeVariable = new AdvanceTypeVariable();
            t.TypeVariable.Name = name;
            return t;
        }

        /// <summary>
        /// Creates a bounded variable type.
        /// </summary>
        /// <param name="name">variable name</param>
        /// <param name="upperBound">True if type bounds are the upper bounds, e.g., {@code T super B1, B2, B3}</param>
        /// <param name="bound1">first bound</param>
        /// <param name="boundsRest">more bounds</param>
        /// <returns>created type</returns>
        public static AdvanceType CreateType(string name, bool upperBound, AdvanceType bound1, params AdvanceType[] boundsRest)
        {
            AdvanceType t = CreateType(name);
            t.TypeVariable.IsUpperBound = upperBound;
            t.TypeVariable.Bounds.Add(bound1);
            t.TypeVariable.Bounds.AddRange(boundsRest);
            return t;
        }

        /// <summary>
        /// Creates a bounded variable type.
        /// </summary>
        /// <param name="name">variable name</param>
        /// <param name="upperBound">True if type bounds are the upper bounds, e.g., {@code T super B1, B2, B3}</param>
        /// <param name="bounds">type bounds</param>
        /// <returns></returns>
        public static AdvanceType CreateType(string name, bool upperBound, ICollection<AdvanceType> bounds)
        {
            AdvanceType t = CreateType(name);
            t.TypeVariable.IsUpperBound = upperBound;
            t.TypeVariable.Bounds.AddRange(bounds);
            return t;
        }

        /// <summary>
        /// Creates a concrete (e.g., no type arguments or parametric type with the supplied base type and type arguments.
        /// </summary>
        /// <example>createType(COLLECTION, createType(STRING)</example>
        /// <param name="baseType">base type</param>
        /// <param name="types">type arguments.</param>
        /// <returns>created new type</returns>
        public static AdvanceType CreateType(Uri baseType, ICollection<AdvanceType> types)
        {
            AdvanceType t = new AdvanceType();
            t.TypeURI = baseType;
            t.TypeArguments = new List<AdvanceType>(types);
            return t;
        }

        /// <summary>
        /// Creates an {@code advance:type} type constructor XML
        /// </summary>
        /// <param name="type">type to convert to XML</param>
        /// <returns>Type Xml</returns>
        public static XmlNode CreateType(AdvanceType type)
        {
            XmlDocument doc = new XmlDocument();
            XmlElement node = doc.CreateElement("type");
            type.FillXmlElement(node);
            return node;
        }

        /// <summary>
        /// Converts the XML type declaration into an actual AdvanceType object.
        /// </summary>
        /// <param name="typeXml">type XML</param>
        /// <returns>Type object</returns>
        public static AdvanceType GetType(XmlNode typeXml)
        {
            return CreateFromXml<AdvanceType>(typeXml);
        }
    }
}
